### PR TITLE
trying to understand Q.combine behavior

### PR DESCRIPTION
**Pull request by cansarigol**
Ticket: [#26368](https://code.djangoproject.com/ticket/26368)

### CODE REVIEW

File: tests/queries/tests.py [hunk 1]
```diff
@@ -3224,8 +3224,10 @@ def test_ticket_21748_complex_filter(self):
         # Finally, a more complex case, one time in a way where each
         # NOT is pushed to lowest level in the boolean tree, and
         # another query where this isn't done.
+        #
+        # ~Q(program__id=p1.id) = NOT ( "queries_program"."id" IS NOT NULL AND "queries_program"."id" = 1 )
         qs1 = Identifier.objects.filter(
-            ~Q(~Q(program__id=p2.id, channel__id=c1.id) & Q(program__id=p1.id))
+            ~Q(~Q(program__id=p2.id, channel__id=c1.id) & Q(program__id=p1.id) & Q(program__id__isnull=False))
         ).order_by('pk')
```
Comment (cansarigol) [inline tests/queries/tests.py]: After changing, this part blew up. When I examined the query, I saw it is wrong. 
just alone `~Q` object => `~Q(program__id=p1.id) = NOT ( "queries_program"."id" IS NOT NULL AND "queries_program"."id" = 1 )` but in before version the qury was like this: `~Q(program__id=p1.id) = NOT ("queries_program"."id" = 1 )`

**Comment by carltongibson** [comment]
Hey @cansarigol. 

It's not instantly clear what's going on here. 

Can you help me by adding a **separate** test case that shows the behaviour from the issue, rather than editing the existing test cases? 

Then we should have one test that fails: the one you add. (The task then is to make that pass without breaking any others.)

Thanks!

**Comment by cansarigol** [comment]
Hi @carltongibson, Actually I used the sample from the ticket. 
```
not_in_age_range = ~Q(position__job__employee__age__range=(0,100))
has_job = Q(employee__job__id__isnull=False)
print(IdealJob.objects.filter(has_job & not_in_age_range).query)
```
[query string](https://dpaste.de/a9am)

the problem is this part: `AND U2."id" =  "app_job"."id"`

these forms don't have any problem:
```
IdealJob.objects.filter(has_job).filter(not_in_age_range).query
IdealJob.objects.filter(not_in_age_range & has_job).query
```

**Comment by carltongibson** [comment]
What I mean is to leave the existing test cases alone, and add a new one, not edit the existing test cases. (That way we can ≈know we didn't break anything. And the diff will be clearer.)

**Comment by cansarigol** [comment]
Hey @carltongibson , I added a [commit](https://github.com/cansarigol/django/commit/0a272c937f1c8b85a31ac6408bb46fa95537cda7), is it like what you want?

**Comment by cansarigol** [comment]
> Hey @carltongibson , I added a [commit](https://github.com/cansarigol/django/commit/0a272c937f1c8b85a31ac6408bb46fa95537cda7), is it like what you want?

Hi @carltongibson, did you have a chance to take a look?

**Comment by auvipy** [comment]
i think this can be closed as stale

**Pull request closed without merging**